include/deal.II/base/mpi_tags.h

```diff
@@ -0,0 +1,99 @@
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mpi_tags_h
+#define dealii_mpi_tags_h
+
+#include <deal.II/base/config.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+namespace Utilities
+{
+  namespace MPI
+  {
+    namespace internal
+    {
+      /**
+       * This enum holds all MPI tags used in collective MPI communications with
+       * using MPI_ANY_SOURCE inside the deal.II library.
+       *
+       * We keep these tags in a central location so that they are unique within
+       * the library. Otherwise, communication that receives packages with
+       * MPI_ANY_SOURCE might pick up packets from a different algorithm.
+       */
+      struct Tags
+      {
+        /**
+         * The enum with the tags.
+         */
+        enum enumeration
+        {
+          /// Utilities::MPI::some_to_some()
+          mpi_some_to_some = 300,
+
+          /// Utilities::MPI::compute_point_to_point_communication_pattern()
+          compute_point_to_point_communication_pattern,
+
+          /// GridTools::exchange_cell_data_to_ghosts():
+          exchange_cell_data_to_ghosts,
+
+          /// Triangulation<dim, spacedim>::communicate_locally_moved_vertices()
+          triangulation_communicate_locally_moved_vertices,
+
+          /// dof_handler_policy.cc: communicate_mg_ghost_cells()
+          dofhandler_communicate_mg_ghost_cells,
+
+          /// dof_handler_policy.cc: communicate_mg_ghost_cells()
+          dofhandler_communicate_mg_ghost_cells_reply,
+
+          /// mg_transfer_internal.cc: fill_copy_indices()
+          mg_transfer_fill_copy_indices,
+
+          /// SparsityTools::sparsity_tools_distribute_sparsity_pattern()
+          sparsity_tools_distribute_sparsity_pattern,
+
+          /// Dictionary::reinit()
+          dictionary_reinit,
+
+          /// ConsensusAlgorithmPayload::get_requesters()
+          consensus_algorithm_payload_get_requesters,
+
+          /// FETools::extrapolate()
+          fe_tools_extrapolate,
+          /// FETools::extrapolate(), allocate space for 10 rounds:
+          fe_tools_extrapolate_end = fe_tools_extrapolate + 10,
+
+          /// ConsensusAlgorithm_NBX::process
+          consensus_algorithm_nbx_process_request,
+          /// ConsensusAlgorithm_NBX::process
+          consensus_algorithm_nbx_process_deliver,
+
+          /// ConsensusAlgorithm_PEX::process
+          consensus_algorithm_pex_process_request,
+          /// ConsensusAlgorithm_PEX::process
+          consensus_algorithm_pex_process_deliver,
+
```

Comment:
I get a kick out of it every time I recall that C allows you to end the last entry of an enum with a comma...

Reply:
not to be confused with the comma operator (why is that one not allowed inside an enum?) :-)

Reply:
The *use* of the comma operator is allowed. But you can't declare member functions/operators in `enum`s.

That said, you can't declare *global* operators for enums, and I suspect you can also declare comma operators for them!

Reply:
To nobody's surprise, I had to try this out:
```
#include <iostream>

enum E
{
      a,b
};


std::string operator, (const E e1, const E e2)
{
  return "surprise";
}


int main ()
{
  std::cout << (a,b) << std::endl;
}
```
This works and prints "surprise". Now, you might ask why I didn't write `std::cout << a,b << std::endl;` without the parentheses -- that runs into trouble with operator precedences :-)